### ISSUE TITLE
Plugin APIs for interacting with universe constraints are confusing

### ISSUE THREAD

**Issue by tlringer** [coq/coq#7669]
#### Version

Coq 8.8

#### Description of the problem

As a plugin developer, I had no idea that using constructors like `mkApp` and so on does not add the proper universe constraints, and that I need to refresh them myself sometimes. It's not clear when I need to do that, and even then, it's not clear which interface to call and how. Right now I call type-checking before every term I define in my plugin, but this can be unnecessarily inefficient. 

It would be great to either abstract this away from plugin developers a bit more, or to clarify when it is necessary to update the `evar_map` and how to best accomplish this. From what I gather, `econstr` should help force better discipline, but I'm not sure that is enough (I never have evars, so right now I still always work in `constr`).

**Comment by ejgallego** [comment]
cc: #7155

**Comment by SkySkimmer** [comment]
When you do `mkApp` you should have some explanation which makes you believe it is well typed. That explanation should tell you if you need any universe constraints. I don't see a way to put that in the interface.
I guess we could tell people to run conversion for each application node but that's still pretty inefficient (also issues may come up with other term constructors, it's not just App that needs to be well typed).

**Comment by ejgallego** [comment]
also cc: #6414

**Comment by tlringer** [comment]
I'm not sure what the best solution is, but if it helps, for a long time I had no idea that the `evar_map` even contained universe constraints. Since I was dealing with terms that never had any evars, I always assumed I could just ignore it. Later on I found out that this was a very incorrect assumption.

**Comment by ejgallego** [comment]
Good point @tlringer , indeed I had the same exact confusion some months ago. @ppedrot suggested the name "unification status", but that also seems a bit confusing. An issue here is that there are two kind of "metas", term evars, and universe variables, so indeed, when creating terms, these have some universe variables by default I do believe.

**Comment by gares** [comment]
> I don't see a way to put that in the interface.

The point is exactly that: with the current model there is no way to get a decent API. My conclusion is that we should question the current design, rather than living with the current API.

In the previous (pre 8.5) model the kernel would infer, and not only check, universe constraints.
At that time tactics were completely ignoring universe constraints.

Two (orthogonal, to my point of view) choices were made:
1. make tactics declare universe constraint and use them (ie. fail if not satisfiable)
2. make the kernel check that the set of constraints declared by tactics are sufficient to type check the term (ie no inference of constraints)

To support the current design I heard the following arguments:
1. if you do HoTT style formalization you really want tactics to understand universes
2. inference is more complex and expensive than checking, so by pulling inference out of the kernel we obtain a safer and faster system

I'm 100% convinced that the first argument is true, while I suspect that the second one is not. First @jhjourdan told me (please confirm of tell me I misunderstood) that the current algorithm is able to perform inference and does it fast. Second, inference is (in some form) still in the kernel since @mattam82 told me there is no way to type check modules/functors without it (again, please confirm this claim or tell me I misunderstood).

What looks reasonable to me is the following design:
- tactics *can* understand universes. Good tactics declare and check constraints
- the kernel *infers* constraints. Eventually checks that the constraints declared by tactics are sufficient (eg emit a warning, or abort if coq is run in -test-mode).

With this design any piece of code not being a tactic can simply ignore they exists. Eg @tlringer plugins could just build a term and ask the kernel to type check it. Also a tactic "forgetting" to declare the right universe constraints would work. Not perfectly if you do HoTT (it may not fail when it should), but *perfectly* for all other users (that is the majority). Eg there is no real use of universes in math-comp, but I had to repair most ssreflect tactics in order to be able to compile that library (and I've not even fished cleaning up the mess IMO). With the design I propose ssr would have just worked out of the box.

**Comment by ppedrot** [comment]
> I'm 100% convinced that the first argument is true, while I suspect that the second one is not. First @jhjourdan told me (please confirm of tell me I misunderstood) that the current algorithm is able to perform inference and does it fast.

Trueish. Indeed, the kernel can perform universe constraint *inference*, but what also needs to be done is universe *unification*. While the two are related, they are not the same, and the latter only occurs outside of the kernel, thanks to all the gods of soundness. I definitely don't want to perform it inside, as it is a mere heuristic, and I don't know if we can accomodate an API that works if one drops universe unification data.

> Second, inference is (in some form) still in the kernel since @mattam82 told me there is no way to type check modules/functors without it (again, please confirm this claim or tell me I misunderstood).

Currently the kernel still performs universe constraint inference for module subtyping, but IMO it's really a wart that should be removed, because it mixes module elaboration (that should be done outside of the kernel) and module checking.

> With this design any piece of code not being a tactic can simply ignore they exists.

Dually to you, I am not convinced that the API simplicity argument is a good one. There is a lot of things going on during typing, and dropping a state just because it happens to work often doesn't mean that you wrote something correct. Having to think twice about the constraints forces you to write more correct code. Also, the same kind of issue could happen if we use a more clever implementation of `EConstr` with evars in general. I think that we just miss a wrapper for end-users that re-performs type-checking at the end of a proof and does all the universe mumbo-jumbo so that it just works.

**Comment by gares** [comment]
> dropping a state

I understand your arguments but who decides what is a state? The fact that universes are in the proof state is a choice that was made. Before they were not.



> think that we just miss a wrapper

So you are saying that since the kernel does not do inference we keep a copy of it that does and we run both in tandem?



> universe unification

Can you elaborate? The pre 8.5 algorithm was handling `Eq` constraints (eg during conversion). To me this sound like unification of two universe variables, so I must miss some detail.

**Comment by SkySkimmer** [comment]
>So you are saying that since the kernel does not do inference we keep a copy of it that does and we run both in tandem?

We already have it, it's `Typing`.

>Can you elaborate? The pre 8.5 algorithm was handling Eq constraints (eg during conversion). To me this sound like unification of two universe variables, so I must miss some detail.

With universe polymorphism, minimization is critical for performance, but it's not something that can reasonably be put in the kernel.

**Comment by gares** [comment]
> We already have it, it's Typing

`typing.ml` works on open terms (they can contain evars) and indeed it uses unification. It is not a kernel + universe constraint inference (that uses conversion). 
And even if it was, type checking a term twice (just before the Qed) *and* at Qed is, believe me, not fast. In math-comp Qed takes 1/3 of the entire time (IIRC this is on par with most scripts using little ltac automation). We can't afford doing it twice. The "wrapper" @ppedrot talks about is never going to happen (at least not in this form) so it is not a valid argument, IMHO.

> With universe polymorphism, minimization is critical for performance, but it's not something that can reasonably be put in the kernel.

I'm sorry I'm not very knowledgeable about minimization as of today but I think you are misunderstanding my proposal. If you do HoTT stuff, you have many universes and you want to use tactics that track constraints. Such constraints can be minimized (and the term rewritten to only contain the minimal set of univ variables) *before* calling the kernel. Exactly as we do today.
Still, if you set of minimal constraints lacks one (because your heuristic is buggy, or because tactics are buggy), the kernel can add one (if you want it to infer) or error (if you want it to check, eg explicit constraints given next to a poly definition or because you run in -test-mode).

**Comment by gares** [comment]
Maybe my intent here is not clear. I'm questioning the current design because I'd like to find a better one.
I invite all of you to play the same, and question the current design together with all the arguments we have all used to support it.

**Comment by ppedrot** [comment]
@gares We should probably summon @mattam82 to this conversation, he designed the universe implementation after all.

**Comment by SkySkimmer** [comment]
For SProp I did end up making the kernel re-infer the relevance marks which the upper layers should in principle produce. Fixing the upper layers would have taken more invasive changes and still left many sprop-unaware functions.

Unlike universe constraints, which we want to postprocess according to some heuristics, there is a unique right answer for what relevance marks should be produced.

If we make the kernel infer constraints, it may be easier to write upper layers but there's a higher risk of them being incorrect and breaking in some cases (mostly when encountering universe polymorphism).
There are also possible issues with added code complexity and decreased performance but I think those wouldn't be blockers.

**Comment by gares** [comment]
> If we make the kernel infer constraints, it may be easier to write upper layers but there's a higher risk of them being incorrect and breaking in some cases (mostly when encountering universe polymorphism).

Please elaborate.

**Comment by mattam82** [comment]
I agree the current API puts more pressure on the clients to be correct, but not much more than being correct w.r.t. evars in general (i.e. the same state passing applies to both contexts). For the case where no evars are involved evidently there is a loss in simplicity but the majority of tactics need to be evar-aware already. I don't see the problem with calling Typing to infer potentially missing constraints before going into the kernel and don't see how allowing the kernel itself to re-infer constraints would simplify the API. In other words, it would just be a matter of doing what Typing is doing inside the kernel, so I'm pretty sure you would end up with the same API in both cases.

I think the simplest methodology is to build terms using EConstr, properly threading the evar_map and calling a wrapper which calls Typing before declaring a definition to infer the constraints. Note that with universe polymorphic definitions, things become a bit more complex because you need to know the "universe"-arities of the constants you define, but following the EConstr way and constructing references to polymorphic references using new_global and such makes it transparent. 

As a more abstract API, we could certainly make an LCF-style interface containing functions for each typing rule/construct which would check and add the constraints incrementally (and do the necessary unifications in presence of evars) so that Typing is not necessary in the end.

**Comment by gares** [comment]
> In other words, it would just be a matter of doing what Typing is doing inside the kernel, so I'm pretty sure you would end up with the same API in both cases.

This is my feeling too (about doing what Typing does, but in the kernel) but I don't agree about the "sameness" kernel API. While the API could have the same types, their meaning would be different. Adding a `term * UContext.t` to a `safe_env` would return a `safe_env` with a new entry (for the term) and potentially more constraints than the ones in the `UContext.t` (I'm sorry if I got the name wrong, maybe it is a `UState.t`). To me thew new API looks simpler (eg passing the empty set of constraints would work). Sure, one could be strict and check the given set of constraints is sufficient, but it would be optional (let's call this *strict mode*). Today it is mandatory.

> calling a wrapper which calls Typing before declaring a definition to infer the constraints.

This argument was raised by @SkySkimmer before and questioned by me in https://github.com/coq/coq/issues/7669#issuecomment-396177260 (first part). I've the feeling that the proposed solution is to type check the term twice. If the kernel can do the inference, then we need to type check only once.

I'm sorry for the lengthy messages, but I'd like to raise one last problem with the current approach. It requires tactics/plugins to guess/know how the kernel will compares terms. We know that the constraints depend on how much we unfold in the kernel but IMO a tactic/plugin cannot/should-not know. The status quo is that the kernel does *backtrack* if the constraints declared by tactics seem not to be sufficient, hoping that unfolding more the terms would require less constraints. If it the kernel was allowed to infer the constraints (eg in non strict mode), it could just avoid backtracking.

**Comment by mattam82** [comment]
Indeed, if you're not doing the work of inferring constraints outside the kernel and go through this API, you would do the check twice.

The kernel will backtrack during inference just like during checking, if only because the user can give universe annotations that have to fire unfolding. If it unfolded always (i.e. always reduce polymorphic constants) it would likely diverge for all we can see. If you just try to implement: if the arguments are convertible then lets unify the universe instances without the possibility to backtrack then you break transitivity of delta-conversion. So, I don't see why you would be in an easier position.

**Comment by mattam82** [comment]
To be clear, I wouldn't mind having the loose mode, but I fear confusion and lurking bugs. The foremost motivation for handling universes at the tactic level was that you would get the proper universe error at the point you made a universe mistake,  and that you could rely on the fact that an evar_map's content would be well typed w.r.t. universes (that seems like a good sanity check) and not wait for Qed. Actually we could maybe even devise an API to pass a trusted evar_map to the kernel to avoid retypechecking at Qed: you basically need to trust only `new_evar` and `define` which does the kernel-side type-checking test of a term fragment (the API with just these two operations ensures monotonicity I believe). Back to the question, strict declarations helped me unveil during development a lot of places where evar_maps were mishandled and typechecking was actually needed. And now it's a basis on which we can handle
parallel checking of universe-polymorphic proofs if the user declare the constraints.

About unification of universes, I believe it could be adapted to the loose scenario, at least when you're not using universe polymorphism: backtrack/unfold if adding an equality/inequality fails when converting polymorphic references, foregoing completely the flexible/rigid distinction in loose mode. In universe polymorphic case, assuming you give to the kernel some term where no universe unifications ever happened, I suspect without minimization it would just blow up generating huge universe instances and constraints. That's mainly due to cumulativity really, typically when you have a polymorphic function `f : forall (A : Type@{i})...` and apply it to some `T : Type@{j}` the generated constraint is just `j <= i`. Assuming i and j are global it's usually fine and the constraint might already be derivable. If `f` is universe polymorphic however, each application will generate one new universe `k` and fresh constraint `j <= k`.

**Comment by gares** [comment]
Since I still feel the design I propose is not clearly understood, I try the converse approach. I change strategy.
Here a very simple description of the design I propose:

*everything is exactly as today*:
- tactics thread the evar map, add to it universe constraints and so on, fails if universe constraints are not OK, etc..
- the evar map is used, at the env of the proof, to minimize and restrict the universe context and shrink the universes present in the proof term

*but*:
- the kernel is allowed to add a constraint if not already present.

To my eyes, what changes is that users don't see bugs by tactics not declaring constraints, plugins not calling Typing, etc.

Let's start from this, and it is up to you to pin point all the problems of this design. I'll then try to defend it.

**Comment by ejgallego** [comment]
As an "outsider" to this API, I have been thinking for a while what would be more convenient; indeed thanks to @tlringer I got some experience on it; so far I think I'd tend to prefer the solution @mattam82 proposes, ideally we would have term-building combinators that can add the constraints for us, so we are not forced to call type inference at the end.

Delaying all the inference to the end / kernel side looks a bit costly to me, and there are of course the other reasons involved.

**Comment by gares** [comment]
> Delaying all the inference to the end / kernel side looks a bit costly to me

I see no reason why the inference of constraints outside the kernel (eg in `Typing` + `Evarconv`) would be faster than the inference inside the kernel (`Typeops` + `Reduction`). Both algorithms would type-check the input term, in a very similar way, actually in an almost identical way if the term is evar closed.

Maybe it is not clear what I mean by checking vs inference (of universe constraints).

If the constraints were a `Set` (1), we are talking about `Set.add (Le i j) csts` v.s. `Set.mem (Le i j) csts`.
The algorithm doing inference would use `add` the one doing checking would do `mem`. So the main difference is that one doing inference has to thread around `csts` since it may be updated.

Today the kernel does `Set.mem`, so if the constraint is not there you get an error.
I propose to go (back) to the  following code being used in the kernel in place of `Set.mem`:
```ocaml
let add_constraint c csts =
  if strict_mode then
    if Set.mem c csts then csts
    else error "bla bla"
  else 
    Set.add c csts
```
FTR, the code before 8.5 was simply `Set.add`.

And by the way, my proposal leaves *most* of the inference in tactics (you say *all the inference*, so maybe you missed this point). Indeed if tactics are correct wrt universes, then whenever the kernel would do
a `Set.add (Le i j) csts` the operation would be a no-op, since the set `csts` would already
contain the constraint (equivalently if `strict_mode=true` then the error would never be reached). 

-- 
(1): it is not exactly a set, since `add` has to also check that the constraint is admissible, but the API is similar in spirit

**Comment by ejgallego** [comment]
Indeed an "efficient" combinator cannot be implemented then the question is kind of moot.

**Comment by mattam82** [comment]
The idea of the interface would precisely to be working with trusted judgments (i.e. a term and its type) instead of just terms to avoid retypechecking existing terms. Ie. mkApp would become Typing.judge_of_apply. But indeed that's a large change to tactic implementations and wouldn't be compatible with term surgery like map_constr_with_binders. 

I think I understand your design proposal, it can be practically useful to avoid unsatisfied constraints anomalies and it makes the system lenient with plugins that provide partially type-checked terms to the kernel while avoiding costly re-typecheckings, and without risking unsoundness. I also agree it should be a relatively minor change to the code: basically call the infer_conv variant instead of check_conv in non-strict mode and gather the constraints. With this solution we can still have the checker not do any inference in the future. I don't have anything better to propose than calling Typing if you're using loose tactics or crafting constrs by hand without looking at universes, and I agree in the current situation, that's an important use case of the kernel.

**Comment by ppedrot** [comment]
@gares Maybe we slightly overestimate the resort to raw terms in tactic writing. In practice, most people are actually using trusted combinators rather that rolling up their own term. The fact that ssreflect does not do this might not be representative of ML-written tactics. In the latter case, it is enough to retypecheck *locally* the pieces of terms that are pushed into the proof term to resynthesize the constraints.

**Comment by gares** [comment]
> With this solution we can still have the checker not do any inference in the future.

I had some sort of migration strategy in mind, but I did not write it here not to make my messages even longer. 

I think that we could "tag" tactics as either "UnivOK" or "UnivLoose". Then the tactic monad (and for once I'll say good things about it) can be instrumented to track which tactics are used in a proof. A boolean is enough, and if all tactics are "UnivOk" then the kernel is run in strict mode. Maybe one day no "standard" tactic will be tagged as "UnivLoose". We could do the same for commands (we need no monad this time).

As a bonus we could provide a (command line) flag to force loose mode. If a user finds a bug "univ not defined" or "univ x is not < than univ y" we give him an easy work around (just pass the flag) and we try to see which tactic claims to be correct wrt universes but is not.

And yes, checkers (our or 3rd party ones) are not supposed to infer, but just check universes since we will be able to store in the .vo files the constrains after our kernel has (eventually) fixed them. This is not changing anything wrt today.

**Comment by gares** [comment]
> most people are actually using trusted combinators rather that rolling up their own term. The fact that ssreflect does not do this might not be representative of ML-written tactics.

In ssr we try to be "safe from a user perspective" and type check things in order to signal errors early.
See my bunch of PRs that restore the safety of apply_type, a combinator used not only by ssr...

What is horrible, I agree, is that there is much code that was once correct and based on `map` and that has never been ported to `fold` after something that was not "a state" became a state and had to be threaded.
And this state includes the universe constraints, of course.

The migration strategy above could apply. `=> ipat` and `/view` would be "UnivOK" while the rest would still be mostly "UnivLoose" :-/

**Comment by SkySkimmer** [comment]
Turns out Typing doesn't check that the instance on a polymorphic constant is of the right length nor that it satisfies the right constraints. So if you're not building your constants through fresh_global or similar, or if you drop the constraints, then just calling typing before handing off to the kernel won't work.
